### DISCUSSION TITLE
Azure DPS Connection

### DISCUSSION THREAD

**Discussion by FHFS** [nanoframework/Home#1561] [Q&A] [answered]
Hi all,

I am having some trouble connecting to Azure DPS using my device certificates.
```
    ++++ Exception System.Net.Sockets.SocketException - CLR_E_FAIL (1) ++++ 
    ++++ Message:  
    ++++ System.Net.Security.SslNative::SecureConnect [IP: 0000] ++++ 
    ++++ System.Net.Security.SslStream::Authenticate [IP: 0054] ++++ 
    ++++ System.Net.Security.SslStream::AuthenticateAsClient [IP: 000a] ++++ 
    ++++ nanoFramework.M2Mqtt.MqttNetworkChannel::Connect [IP: 0063] ++++ 
    ++++ nanoFramework.Azure.Devices.Provisioning.Client.ProvisioningDeviceClient::.ctor [IP: 008f] ++++ 
    ++++ nanoFramework.Azure.Devices.Provisioning.Client.ProvisioningDeviceClient::Create [IP: 000a] ++++ 
    ++++ NFBeast.DpsService::Create [IP: 0028] ++++ 
    ++++ Program::<Main>$ [IP: 00f1] ++++ 

    ++++ Exception nanoFramework.M2Mqtt.Exceptions.MqttConnectionException - 0x00000000 (1) ++++ 
    ++++ Message: Socket exception connecting to the broker: -1 
    ++++ nanoFramework.M2Mqtt.MqttClient::Connect [IP: 0049] ++++ 
    ++++ nanoFramework.Azure.Devices.Provisioning.Client.ProvisioningDeviceClient::.ctor [IP: 008f] ++++ 
    ++++ nanoFramework.Azure.Devices.Provisioning.Client.ProvisioningDeviceClient::Create [IP: 000a] ++++ 
    ++++ NFBeast.DpsService::Create [IP: 0028] ++++ 
    ++++ Program::<Main>$ [IP: 00f1] ++++ 
Exception thrown: 'nanoFramework.M2Mqtt.Exceptions.MqttConnectionException' in nanoFramework.M2Mqtt.dll 
An unhandled exception of type 'nanoFramework.M2Mqtt.Exceptions.MqttConnectionException' occurred in nanoFramework.M2Mqtt.dll 
Additional information: Socket exception connecting to the broker: -1 
```

``` 
            X509Certificate azureCA = new X509Certificate(Resource.GetString(Resource.StringResources.AzureRootCerts)); 
            X509Certificate2 deviceCert = new X509Certificate2(chain, privateKey, ""); 
            var provisioning = ProvisioningDeviceClient.Create(DpsAddress, IdScope, DeviceID, deviceCert, azureCA); 
```
I am not sure what is going wrong. 
I have used this same public/private key pair on my esp32 esp-idf project before, without issue.
I have tried the same Root certificate as on my esp32 esp-idf project, the one listed in [nanoFramework.Azure.Devices](https://github.com/nanoframework/nanoFramework.Azure.Devices), and one I got by using openssl showcert.

I could use some help in narrowing down the problem.

**Comment by alberk8** [comment]
First what ESP32 module are you using?
Second is your Certificate dates are valid?.

**Comment by FHFS** [comment]
I am using the ESP32-S3
Yes the Certificates are valid I use them in my esp-idf project to connect to Azure with DPS.

**Reply by alberk8** [reply]
very likely there is not enough memory to run the MQTT with the TLS certificates.

**Reply by FHFS** [reply]
Do the ESP32_S3_ALL and ESP32_S3_BLE nF Firmware images not use the PSRAM on the esp32s3?

**Reply by alberk8** [reply, marked as answer]
It all depends on the module that you have, some modules are without PSRAM.  Then S3 have two version of PSRAM (Quad and Octa speed) and it needs the correct Firmware to activated it.

**Reply by FHFS** [reply]
Got it.  I had to use "ESP32_S3_ALL". As @alberk8 suggests this firmware probably does not support the PSRAM on my device.

Thanks for the support alberk8!

**Reply by FHFS** [reply]
> Then S3 have two version of PSRAM (Quad and Octa speed) and it needs the correct Firmware to activated it.

I have the module with PSRAM, but do not know if it is the Quad or Octa. I'll go figure that out. Do you have the names of the ready to deploy firmware images for Quad and Octa, and where I can find them? I didnt see this info on the esp32s3 documentation page.

**Reply by alberk8** [reply]
Check here for the different modules https://www.espressif.com/en/products/modules   You need to dig in more info in the specification datasheets.

**Reply by FHFS** [reply]
Do you know where I can find the build configuration for the prebuilt nanoFramework firmware images "ESP32_S3" "ESP32_S3_ALL" "ESP32_S3_BLE"?

**Reply by Ellerbach** [reply]
> Do you know where I can find the build configuration for the prebuilt nanoFramework firmware images "ESP32_S3" "ESP32_S3_ALL" "ESP32_S3_BLE"?

Here: https://github.com/nanoframework/nf-interpreter/blob/main/targets/ESP32/CMakePresets.json

And follow the different includes in the first lines if you need something specific.

**Reply by FHFS** [reply]
Thanks!

**Reply by FHFS** [reply]
Further information for future reference.

I was using the ESP-Thread-BR device with a "ESP32-S3 and 2MB Quad SPI Embedded PSRAM". Using "ESP32_S3_BLE" on this device I was able to connect to Azure DPS and IoTHub without issue.

I also have a ESP32-S3-ETH board with a "ESP32-S3 and 8MB Octal SPI Embedded PSRAM". Using "ESP32_S3_ALL" on this device I was able to connect to Azure DPS and IoTHub without issue.

The readme at [nf-interpreter - esp32s3](https://github.com/nanoframework/nf-interpreter/tree/429ba2c8eb218886bad4e89e16d8b367fa89b566?tab=readme-ov-file#esp32_s3-boards) specifies quad or octal support.
